### PR TITLE
Allow for granular section restyling via  convenience api

### PR DESCRIPTION
**Pull request by timkpaine**
# Summary

This is a work-in-progress PR for allowing more granular styling of the various components via the API currently used for cell styling (e.g. `loc.body()`). The motivation is that it's often desired to style e.g. all column labels at once. It looks like this was already started in part (there are many TODOs such as [these](https://github.com/posit-dev/great-tables/blob/7eae1f269a7d1f787de7a2c9198d775933794fd5/great_tables/_utils_render_html.py#L103)). 

This PR is nowhere near done yet, so please don't approve CI running (it will just make noise).

Early Example:
```python
from great_tables import GT, style, loc
from superstore import superstore

(
    GT(superstore(10))
    .tab_style(style=style.text(color="red"), locations=loc.column_labels())
)
```

Screenshot 2024-05-14 at 13 42 14

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/).
- [x] I have followed the [Style Guide for Python Code](https://peps.python.org/pep-0008/) as best as possible for the submitted code.
- [ ] I have added **pytest** unit tests for any new functionality.

### CODE REVIEW

**Comment by timkpaine** [comment]
Screenshot 2024-05-14 at 13 53 17

**Comment by timkpaine** [comment]
Screenshot 2024-05-14 at 14 09 07

File: great_tables/_locations.py [hunk 1]
```diff
@@ -417,11 +566,46 @@ def _(loc: LocBody, data: GTData, style: list[CellStyle]) -> GTData:
     for col_pos in positions:
         row_styles = [entry._from_row(data._tbl_data, col_pos.row) for entry in style_ready]
         crnt_info = StyleInfo(
-            locname="data", locnum=5, colname=col_pos.colname, rownum=col_pos.row, styles=row_styles
+            locname="cell", locnum=5, colname=col_pos.colname, rownum=col_pos.row, styles=row_styles
         )
         all_info.append(crnt_info)
 
-    return data._replace(_styles=data._styles + all_info)
+    data._styles.extend(all_info)
```
Comment (machow) [inline great_tables/_locations.py:573]: I think we may need to convert every use of `data._styles.extend(...)` or `data._styles.append(...)` to `data._replace(...)`.

The extend and append methods mutate the styles list, which can cause style changes to apply to unanticipated places. (`._replace()` does a shallow copy so should be fast to do)

**Comment by machow** [comment]
Hey, thanks for this incredibly comprehensive PR! At first I was nervous about the idea of a PR implementing all locations at once, but after seeing your changes, it seems very reasonable! Happy to kick the tires, help with anything :)

File: great_tables/_locations.py (hunk 1 above)
Comment (timkpaine) [reply great_tables/_locations.py:573]: Ok I went the other way, but I'll revert back!

**Comment by machow** [comment]
Here's my quick attempt at mapping the names here, to those used in the R library gt:

<details>
<summary>code</summary>

```python
from great_tables import GT
import polars as pl

# I made this in a spreadsheet, then used pl.DataFrame.to_init_repr() :)
df = pl.DataFrame(
    [
        pl.Series("Location category", ['Header', 'Stubhead', 'Column labels', None, 'Stub elements', None, None, 'Body elements', None, 'Footer elements', None, None, None], dtype=pl.String),
        pl.Series("gt R package", ['cells_title', 'cells_stubhead', 'cells_column_labels', 'cells_column_spanners', 'cells_stub', 'cells_row_groups', 'cells_stub_summary', 'cells_body', 'cells_summary', 'cells_footnotes', 'cells_source_notes', 'cells_grand_summary', 'cells_stub_grand_summary'], dtype=pl.String),
        pl.Series("Great Tables", ['LocHeader', 'LocStubhead', 'LocColumnLabels*', 'LocSpannerLabel', 'LocStub', 'LocRowGroupLabel', 'LocSummaryLabel', 'LocBody', 'LocSummary', 'LocFootnotes', 'LocSourceNotes', None, None], dtype=pl.String),
        pl.Series("extra1", ['LocTitle', 'LocStubheadLabel', 'LocColumnLabel', None, 'LocRowLabel', None, None, None, None, 'LocFooter', None, None, None], dtype=pl.String),
        pl.Series("extra2", ['LocSubtitle', None, None, None, None, None, None, None, None, None, None, None, None], dtype=pl.String),
    ]
)
GT(df).sub_missing(missing_text="")
```

</details>

image

It seems like you may have surfaced two important things:

1. The R program sometimes maps to sub-locations. E.g. cells_title(groups=...) -> title or subtitle.
2. If you just represent the sub-locations directly (e.g. your use of LocTitle and LocSubtitle), then they can be used directly. This seems like a nice simplification :).

### Potentially replacing `StyleInfo(loc="some_string")`

It seems like defining all Loc sub-locations might also let us replace things like `StyleInfo(loc="footnotes")` with `StyleInfo(loc=loc)`, where `loc` is the LocFootnotes object itself. Replacing all the string checks with isinstance would make it very easy to find references to Loc classes in VS Code, etc..!

In any event, this is looking really great! Thanks for this huge effort---I'm happy to check anything, do anything to support!

File: great_tables/_utils_render_html.py [hunk 2]
```diff
@@ -341,50 +317,31 @@ def create_columns_component_h(data: GTData) -> str:
             table_col_headings = tags.tr(level_1_spanners, class_="gt_col_headings gt_spanner_row")
 
     if _get_spanners_matrix_height(data=data) > 2:
```
Comment (timkpaine) [inline great_tables/_utils_render_html.py:319]: move to separate PR?
Screenshot 2024-05-14 at 15 16 24
Screenshot 2024-05-14 at 15 16 17

Comment (machow) [reply great_tables/_utils_render_html.py:319]: I would definitely punt this piece to a new PR (or I can take it up). I just tested and spanners above 1 extra level (e.g. a spanner over columns) don't work currently...

```python
from great_tables import GT, exibble

GT(exibble).tab_spanner("A", ["num", "char"]).tab_spanner("B", ["fctr"]).tab_spanner("C", ["A", "B"])
```

Realistically, this code should be raising a NotImplementedError afaict

Comment (machow) [reply great_tables/_utils_render_html.py:319]: Ah wait, I'm realizing you fixed the spanner implemention! For sure feel free to punt to a separate PR, but realistically you're the hero we all need so if you to keep it in here happy to test / validate in this PR (if more convenient)

Comment (timkpaine) [reply great_tables/_utils_render_html.py:319]: It's not quite correct, im going to breakout and finish in a separate PR

Comment (timkpaine) [reply great_tables/_utils_render_html.py:319]: https://github.com/posit-dev/great-tables/pull/345

**Comment by machow** [comment]
I'm going to try and complete this PR this week, if that's okay! It's looking really good, and I think will be nice to have for WIP on rendering GT objects interactively (in [reactable-py](https://github.com/machow/reactable-py)) 😁

**Comment by timkpaine** [comment]
🚀

File: great_tables/_gt_data.py [hunk 3]
```diff
@@ -610,7 +611,7 @@ def order_groups(self, group_order: RowGroups):
         # TODO: validate
         return self.__class__(self.rows, self.group_rows.reorder(group_order))
 
-    def group_indices_map(self) -> list[tuple[int, str | None]]:
+    def group_indices_map(self) -> list[tuple[int, GroupRowInfo | None]]:
```
Comment (machow) [inline great_tables/_gt_data.py:614]: now returning GroupRowInfo directly, since we need the group ids, not just their defaulted labels to match group name locs

File: great_tables/_gt_data.py [hunk 4]
```diff
@@ -869,8 +870,7 @@ class FootnoteInfo:
 
 @dataclass(frozen=True)
 class StyleInfo:
-    locname: str
-    locnum: int
+    locname: Loc
```
Comment (machow) [inline great_tables/_gt_data.py:873]: locname with a Loc class now fully identifies a style info, so no need for locnum

File: great_tables/_locations.py [hunk 5]
```diff
@@ -383,27 +417,114 @@ def _(loc: LocBody, data: GTData) -> list[CellPos]:
...
+# LocSourceNotes
+
+
 @singledispatch
 def set_style(loc: Loc, data: GTData, style: list[str]) -> GTData:
     """Set style for location."""
     raise NotImplementedError(f"Unsupported location type: {type(loc)}")
 
 
+@set_style.register(LocHeader)
```
Comment (machow) [inline great_tables/_locations.py:445]: Since we removed .locnum and just use the Loc isntances directly, we can consolidate most simple versions of `set_style()`

File: great_tables/_locations.py [hunk 6]
```diff
@@ -355,15 +366,38 @@ def resolve(loc: Loc, *args: Any, **kwargs: Any) -> Loc | list[CellPos]:
...
+    cols = resolve_cols_i(data=data, expr=loc.columns)
+    cell_pos = [CellPos(col[1], 0, colname=col[0]) for col in cols]
+    return cell_pos
+
+
+@resolve.register
+def _(loc: LocRowGroups, data: GTData) -> set[int]:
+    # TODO: what are the rules for matching row groups?
+    # TODO: resolve_rows_i will match a list expr to row names (not group names)
+    group_pos = set(pos for _, pos in resolve_rows_i(data, loc.rows))
```
Comment (machow) [inline great_tables/_locations.py:391]: Note that this currently can't match row groups by their name (which is also their id). It can only match by index number or match every group.

File: great_tables/_gt_data.py (hunk 4 above)
Comment (rich-iannone) [reply great_tables/_gt_data.py:873]: Great!

File: great_tables/_locations.py (hunk 6 above)
Comment (rich-iannone) [reply great_tables/_locations.py:391]: Totally okay since we don't currently make it possible for the user to set nice IDs anyhow.

**Review by rich-iannone — APPROVED** [review]
LGTM!
